### ISSUE TITLE
Why do you need Packaged COM?

### ISSUE THREAD

**Issue by DrusTheAxe** [roxk/packaged-oop-winrt-server-app-extension#2]
TL;DR Use WinRT + appextension + Dynamic Dependencies. No need for Packaged COM.

Neat example. But if you want to access WinRT servers there's an easier answer than Packaged COM. This statement is the key:

>...there is no trivial way to locate the binary which would handle the activation request even though an app had declared it in the manifest.

There is - use AppExtension. 

The recommended solution is to use **AppExtension** + Dynamic Dependencies for **Discovery** + Access

* WinRT provider
  * Create MSIX package containing your WinRT implementation
  * Declare `<Extension Category="windows.appExtension"><uap3:AppExtension Name="com.contoso.pixeleditor.extension.filter"...>` in your `appxmanifest.xml` to advertise 'this package has filter extensions for Contoso's PixelEditor'
* WinRT consumer
  * Use `AppExtensionCatalog.Open("com.graphicco.painteditor.extension.filter").FindAllAsync() to **discover** which packages have extensions you're interested in
  * Use Dynamic Dependencies to add those packages to your process' package graph so you can **access** their content (`ActivateInstance()`, `LoadLibrary()`, etc).

No need to fiddle with CLSIDs and other classic COM-isms :-)

P.S. And given the appextension definition you can provide additional metadata if you desire (e.g. Id, DisplayName, more complex data via Properties).

Links:
* uap3:AppExtension -- https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-uap3-appextension-manual
* AppExtensionCatalog API -- https://learn.microsoft.com/en-us/uwp/api/windows.applicationmodel.appextensions.appextensioncatalog?view=winrt-22621
* Dynamic Dependencies -- https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/framework-packages/use-the-dynamic-dependency-api

**Comment by roxk** [comment]
I feel like I should pin this issue in README😀 Thanks @DrusTheAxe for the comment.

@asklar has kindly helped sharing this sample with internal folks, and one thing concerns them - down level support. DynDep for WinRT server access only works for "normal" apps distributed by ordinary (non-MS) developers if their package had declared `uap15:DependencyTarget` as [documented here](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/framework-packages/use-the-dynamic-dependency-api). Is that true?

If that is true, its very likely that internal teams cannot adopt packaged WinRT server since apparently the teams (at least Dev Home's team) want to support windows 10 as well. On windows 10, app developers wouldn't be able to use `uap15::DependencyTarget`, nor could they distribute framework package on the store. If there are no other ways to allow Widget host and Dev Home app to gain access to developer's WinRT server, the teams would have to fallback to packaged COM...🥲 It hurts.

Any chance of making `uap15:DependencyTarget` work down level, or allowing developers to distribute framework package on the store?

**Comment by DrusTheAxe** [comment]
>down level support

'Downlevel' has multiple meanings in our circles (older Win11, Win10, RS5, ...). Backporting the Win11 support isn't viable and WinAppSDK DynDep doesn't support targeting a Main package but maybe... The hard nut is package-in-use and lifetime management (Framework+Main packages have significant differences here). Supporting Main packages wasn't a requirement and it's (somewhat) different work so it was out-of-scope. Hmmm.

I'll follow up with @asklar offline to better understand the scenario and see what our options might be. Thanks!

>as [documented here](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/framework-packages/use-the-dynamic-dependency-api). Is that true?

Not quite. That says

_Also—as of Windows 11, version 22H2 (10.0; Build 22621)—you can **reference and use main packages** as well. **The main package must** correctly configure its app package manifest source file (the Package.appxmanifest file in Visual Studio). You need to **set `<uap15:DependencyTarget>true</>`** (see [uap15:DependencyTarget](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-uap15-dependencytarget))._

Framework packages can always be targeted by Dynamic Dependencies. `uap15::DependencyTarget` is to enable DynDep to **target** a Main package. IOW the Main package has to opt-in to allow it to be used as a dynamic dependency (frameworks always implicitly allow this).

I guess the docs weren't quite clear on that point. I'll pass along word to improve that. Thanks for bringing this to our attention.

So frameworks are always good to go for dynamic use. But..._discovery_ is an issue. `<Extension Category="windows.appExtension">` is an application-scope extension, thus it can only be declared under an `<Application>`. Frameworks can't specify `<Application>`. The solution is an equivalent `windows.packageExtension` with package scope but that's in our backlog and not available today. There's more info [here](https://github.com/microsoft/WindowsAppSDK/issues/226#issuecomment-705781319). If this interests you please share your feedback.

>DynDep for WinRT server access only works for "normal" apps distributed

That's a bit of a misnomer. DynDep works on **packages**, not apps. It's unclear what you mean by a "normal" app. Please clarify.

**Comment by roxk** [comment]
> That's a bit of a misnomer. DynDep works on packages, not apps. It's unclear what you mean by a "normal" app. Please clarify.

My bad. Just replace the word "app" with "package" and that would be my question:

> DynDep for WinRT server access only works for "normal" (main) package distributed by ordinary (non-MS) developers if their package had declared uap15:DependencyTarget as [documented here](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/framework-packages/use-the-dynamic-dependency-api). Is that true?

From your previous post, I think the answer to this question is a clear yes.

It's also clear that even if developers can distribute framework package, there is no way for developers to declare an extension in appxmanifest for Widget Host/Dev Home to pick up🥲So using framework package for widget/dev home extension is out the window.

> I'll follow up with @asklar offline to better understand the scenario and see what our options might be. Thanks!

Thanks! My mission is completed now that the owners are working on it. Look forward to the result. Hope we can see WinRT server based API with dev home/widget working on 1809+ some days.
